sweep/sweeper.go

```diff
@@ -752,12 +752,86 @@ func (s *UtxoSweeper) bucketForFeeRate(
 }
 
 // createInputClusters creates a list of input clusters from the set of pending
-// inputs known by the UtxoSweeper.
+// inputs known by the UtxoSweeper. It clusters inputs by
+// 1) Required tx locktime
+// 2) Similar fee rates
 func (s *UtxoSweeper) createInputClusters() []inputCluster {
 	inputs := s.pendingInputs
 
-	feeClusters := s.clusterBySweepFeeRate(inputs)
-	return feeClusters
+	// We start by getting the inputs clusters by locktime. Since the
+	// inputs commit to the locktime, they can only be clustered together
+	// if the locktime is equal.
+	lockTimeClusters, nonLockTimeInputs := s.clusterByLockTime(inputs)
+
+	// Cluster the the remaining inputs by sweep fee rate.
+	feeClusters := s.clusterBySweepFeeRate(nonLockTimeInputs)
+
+	// Since the inputs that we clustered by fee rate don't commit to a
+	// specific locktime, we can try to merge a locktime cluster with a fee
+	// cluster.
+	return zipClusters(lockTimeClusters, feeClusters)
+}
+
+// clusterByLockTime takes the given set of pending inputs and clusters those
+// with equal locktime together. Each cluster contains a sweep fee rate, which
+// is determined by calculating the average fee rate of all inputs within that
+// cluster. In addition to the created clusters, inputs that did not specify a
+// required lock time are returned.
+func (s *UtxoSweeper) clusterByLockTime(inputs pendingInputs) ([]inputCluster,
+	pendingInputs) {
+
+	locktimes := make(map[uint32]pendingInputs)
+	inputFeeRates := make(map[wire.OutPoint]chainfee.SatPerKWeight)
+	rem := make(pendingInputs)
+
+	// Go through all inputs and check if they require a certain locktime.
+	for op, input := range inputs {
+		lt, ok := input.RequiredLockTime()
+		if !ok {
+			rem[op] = input
+			continue
+		}
+
+		// Check if we already have inputs with this locktime.
+		p, ok := locktimes[lt]
+		if !ok {
+			p = make(pendingInputs)
+		}
+
+		p[op] = input
+		locktimes[lt] = p
+
+		// We also get the preferred fee rate for this input.
+		feeRate, err := s.feeRateForPreference(input.params.Fee)
+		if err != nil {
+			log.Warnf("Skipping input %v: %v", op, err)
+			continue
+		}
+
+		input.lastFeeRate = feeRate
+		inputFeeRates[op] = feeRate
+	}
+
+	// We'll then determine the sweep fee rate for each set of inputs by
+	// calculating the average fee rate of the inputs within each set.
+	inputClusters := make([]inputCluster, 0, len(locktimes))
+	for lt, inputs := range locktimes {
+		lt := lt
+
+		var sweepFeeRate chainfee.SatPerKWeight
+		for op := range inputs {
+			sweepFeeRate += inputFeeRates[op]
+		}
+
+		sweepFeeRate /= chainfee.SatPerKWeight(len(inputs))
```

Comment:
doesn’t this need to be a weighted average based on the inputs’ actual weight? this method assumes all inputs are of equal weight

Reply:
I think this is correct (if we want to average the feerates), since we taking the average fee rate here, not the fee.  This is pre-existing though, I'm just using the same logic as below.

Reply:
hmm i'm not so sure, say you have two inputs A (100w, 1000sat/kw) and B (200w, 3000sat/kw). individually, input A would pay 100w*1000sat/kw = 100 sat and B would pay 200w*3000sat/kw = 600sat. combined they pay 700sat total for 300w, or 2.333sat/w.

the unweighted average of the fee rates gives you 2000sat/kw, so total of 300w * 2000sat/kw = 600sat.

however, a weighted average produces (1000sat/kw + 2*3000sat/kw)/3 = 2333sat/kw which gives us the expected fee rate. then 100w*2333sat/kw + 200w*2333sat/kw = 700sat as expected.

> This is pre-existing though, I'm just using the same logic as below.

This may be, but I still think it's incorrect :P

Reply:
> which gives us the expected fee rate

Sure, but I don't think it is defined what is the "expected" fee rate here. Here we cluster some inputs with different fee rates togheter, and need to decide which fee rate to use for all of them. Even though one is heavier than the other, I'm not sure it is "more correct" to choose a fee rate that is closer to that one.

I do agree though, that it might be a better strategy. Since many tiny outputs could bring the fee rate down (and since they are tiny they don't cost much anyway) it could be more correct to take the weighted average. You okay with doing this as a follow up (must change it in the pre-existing case as well)?

Reply:
okay, maybe it helps to clarify the objective. say you have the weights and fee rate for each input, `(w_i, r_i), i in [1, n]`. my assumption was that this is trying to compute `r_agg` where `sum (w_i * r_i) = r * sum (w_i)`, where `r_agg` represents the _actual_ fee rate of the cluster after aggregating all the desired fees paid by individual inputs.

a different example, it might make sense to take `r_max = max(r_i)`, which then forces all lower fee-rate inputs to pay the same fee-rate as the input with highest priority. unlike `r_agg`, this way creates a `sweepFeeRate` that never decrease the priority of an input, which seems useful if certain inputs are time-sensitive.

i could also see e.g taking the most recent `r_i`, which is maybe viewed as the most up-to-date fee rate from our estimator/user, and forcing all inputs to readjust to that. this value could fluctuate though, so maybe it's not ideal for time-sensitive inputs.

i agree that there are different heuristics we can choose here, and one may not necessarily be more correct. if the goal is simply to pick an arbitrary value between `min(r_i)` and `max(r_i)`, then an unweighted average works. however it still doesn't translate into anything meaningful, which is why imo it seems less correct than some of the other options available.

> You okay with me doing this as a follow up (must change it in the pre-existing case as well)?

Sure! Yeah I'm okay with a follow up :)

Reply:
Issue: https://github.com/lightningnetwork/lnd/issues/4812